src/compiler/checker.ts

```diff
@@ -9782,15 +9782,15 @@ namespace ts {
             return type.flags & TypeFlags.TypeParameter && !getConstraintFromTypeParameter(<TypeParameter>type);
         }
 
-        function isTypeReferenceWithGenericArguments(type: Type) {
-            return getObjectFlags(type) & ObjectFlags.Reference && some((<TypeReference>type).typeArguments, isUnconstrainedTypeParameter);
+        function isTypeReferenceWithGenericArguments(type: Type): type is TypeReference {

```

Comment:
Technically this is not correct. A type might be a type reference even if this function returns false.

Reply:
Oops, I forgot that 'type is TypeReference' implies that type is *not* TypeReference if the return value is false.